src/els_code_action_provider.erl

```diff
@@ -0,0 +1,128 @@
+-module(els_code_action_provider).
+
+-behaviour(els_provider).
+
+-export([ handle_request/2
+        , is_enabled/0
+        , options/0
+        ]).
+
+-include("erlang_ls.hrl").
+
+%%==============================================================================
+%% els_provider functions
+%%==============================================================================
+
+-spec is_enabled() -> boolean().
+is_enabled() -> true.
+
+%% /**
+%%  * The server provides code actions. The `CodeActionOptions` return type is
+%%  * only valid if the client signals code action literal support via the
+%%  * property `textDocument.codeAction.codeActionLiteralSupport`.
+%%  */
+%% codeActionProvider?: boolean | CodeActionOptions;
+
+-spec options() -> boolean() | map().
+options() ->
+  Capabilities = els_config:get(capabilities),
+  lager:info("code_actions: [Capabilities=~p]", [Capabilities]),
+  true.
+
+-spec handle_request(any(), els_provider:state()) ->
+  {any(), els_provider:state()}.
+handle_request({document_codeaction, Params}, State) ->
+  #{ <<"textDocument">> := #{ <<"uri">> := Uri}
+   , <<"range">>   := RangeLSP
+   , <<"context">> := Context } = Params,
+  Result = code_actions(Uri, RangeLSP, Context),
+  {Result, State}.
+
+%%==============================================================================
+%% Internal Functions
+%%==============================================================================
+
+
+%% result: (Command | CodeAction)[] | null where CodeAction is defined
+%% as follows:
+-spec code_actions(uri(), range(), code_action_context()) -> [map()].
+code_actions(Uri, Range, Context) ->
+  lager:info("code_actions: [Context=~p]", [Context]),
+  #{ <<"start">> := #{ <<"character">> := _StartCol
+                     , <<"line">>      := StartLine }
+   , <<"end">> := _End
+   } = Range,
+  #{ <<"diagnostics">> := Diagnostics } = Context,
+  %% TODO: process optional "only" field in Context
+  Actions0 = [ make_code_action(Uri, C) || C <- Diagnostics],
+
+  Actions = case application:get_env(erlang_ls, test_code_action) of
```

Comment:
I would remove this from the current version.

Reply:
I added it as a template to other devs for how to use this, and to ensure it works via tests.

Perhaps we should remove it as soon as we put in a real code action? If so, I will create an issue.